src/crimson/os/seastore/cached_extent.cc

```diff
@@ -180,4 +180,100 @@ std::ostream &operator<<(std::ostream &out, const lba_pin_list_t &rhs)
   return out << ']';
 }
 
+void BufferSpace::add_buffer(extent_len_t offset, ceph::bufferptr&& ptr)
+{
+  ceph::bufferlist buffer;
+  buffer.append(std::move(ptr));
+  const extent_len_t length_added = buffer.length();
+  const extent_len_t tail = offset + length_added;
+  auto i = buffer_map.find(tail);
+  if (i != buffer_map.end()) {
+    buffer.append(*i->second.release());
+    buffer_map.erase(i);
+  }
+  i = buffer_map.lower_bound(offset);
+  ceph_assert(i == buffer_map.end() || i->first > tail);
+  if (i != buffer_map.begin()) {
+    --i;
+  }
+  if (i != buffer_map.end() &&
+    i->first + i->second->length() == offset) {
+    i->second->append(buffer);
+  }
+  else {
+    buffer_map[offset].reset(new bufferlist(std::move(buffer)));
+  }
+}
+
+ceph::bufferlist BufferSpace::get_data(extent_len_t offset, extent_len_t length)
+{
+  auto i = buffer_map.upper_bound(offset);
+  --i;
+  ceph::bufferlist res;
+  res.substr_of(*i->second.get(), offset - i->first, length);
+  return res;
+}
+
+ceph::bufferptr BufferSpace::build_ptr(extent_len_t length)
+{
+  assert(buffer_map.size() == 1);
+  auto it = buffer_map[0].get();
+  if(!it->is_contiguous()) {
+    it->rebuild();
+  }
+  ceph::bufferptr ptr(it->buffers().front());
+  ptr.set_length(length);
+  buffer_map.clear();
+  return ptr;
+}
+
+bool BufferSpace::check_buffer(extent_len_t offset, extent_len_t length)
+{
+  auto i = buffer_map.upper_bound(offset);
+  if (i == buffer_map.begin()) {
+    return false;
+  }
+  --i;
+  if (i->first + i->second->length() < offset + length) {
+    return false;
+  }
+  return true;
+}
+
+region_list_t BufferSpace::read_buffer(extent_len_t offset, extent_len_t length)
```

Comment:
The name `read_buffer` might be a little bit misleading since the function of this method is to find out the holes within the range `offset~length`

Reply:
Got it. Renamed to get_unloaded_intervals